### PR TITLE
Disallow Nested on transform parameters

### PR DESCRIPTION
**Pull request by wolfs**
Currently, using `@Nested` for transform parameters is pretty complicated, since the managed type needs to be generated separately. We want to disallow `@Nested` on transform parameters for now and enable them again when the support is better.

### CODE REVIEW

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy [hunk 1]
```diff
@@ -188,6 +203,9 @@ class ArtifactTransformValuesInjectionIntegrationTest extends AbstractDependency
         failure.assertHasCause("Property 'missingInput' does not have a value specified.")
         failure.assertHasCause("Property 'absolutePathSensitivity' is declared to be sensitive to absolute paths. This is not allowed for cacheable transforms.")
         failure.assertHasCause("Property 'noPathSensitivity' is declared without path sensitivity. Properties of cacheable transforms must declare their path sensitivity.")
+        failure.assertHasCause("Property 'nested.outputDirectory' is annotated with unsupported annotation @OutputDirectory.")
+        failure.assertHasCause("Property 'nested.inputFile' is declared without path sensitivity. Properties of cacheable transforms must declare their path sensitivity.")
+        errorOutput.count("> Property") == 2 * 7
```
Comment (lptr) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy]: Where does this come from? Can we make it easier to understand by a comment or via some other way?

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy [hunk 2]
```diff
@@ -140,9 +140,22 @@ class ArtifactTransformValuesInjectionIntegrationTest extends AbstractDependency
...
+            }
+                        
+            allprojects {
+                ext.nested = objects.newInstance(NestedType)
+            }
+        """
         setupBuildWithColorTransform {
             params("""
                 extension = 'green'
+                nested.set(project.ext.nested)
```
Comment (lptr) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy]: Hm, is this how nested properties are to be used? Why don't we do this automatically?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformer.java [hunk 3]
```diff
@@ -169,7 +169,10 @@ public void visitDependencies(TaskDependencyResolveContext context) {
             parameterPropertyWalker.visitProperties(parameterObject, ParameterValidationContext.NOOP, new PropertyVisitor.Adapter() {
                 @Override
                 public void visitInputFileProperty(String propertyName, boolean optional, boolean skipWhenEmpty, @Nullable Class<? extends FileNormalizer> fileNormalizer, PropertyValue value, InputFilePropertyType filePropertyType) {
-                    context.maybeAdd(value.call());
+                    Object actualValue = value.call();
+                    if (actualValue != null) {
+                        context.maybeAdd(actualValue);
+                    }
```
Comment (lptr) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformer.java]: Why is this change necessary?

Comment (wolfs) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformer.java]: The added test exposed a bug when the file property is not set (i.e. `.call` returns null`). Then null is passed to `maybeAdd` and the build fails when trying to determine the dependencies of `null`.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy]: What shall we do automatically? Create the nested bean automatically from the interface type? That is mostly not what you want - given that the nested bean probably comes from somewhere else.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy]: I added a new assertion method which should make it clear.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy (hunk 2 above)
Comment (lptr) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy]: Seems like in at least some cases that's what you'd want. But if we don't want to support interfaces here, why don't we use a concrete class instead?

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy [hunk 4]
```diff
@@ -658,4 +661,16 @@ abstract class MakeGreen implements TransformAction<TransformParameters.None> {
...
+        int count = 0
+        validationErrors.each { propertyName, errorMessageOrMessages ->
+            def errorMessages = errorMessageOrMessages instanceof Iterable ? [*errorMessageOrMessages] : [errorMessageOrMessages]
+            errorMessages.each { errorMessage ->
+                count++
+                failure.assertHasCause("Property '${propertyName}' ${errorMessage}.")
+            }
+        }
+        assert errorOutput.count("> Property") == count
+    }
```
Comment (lptr) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy]: 👍

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy]: It needs to be a managed type, so it needs to be generated by Gradle from an interface. If the type is not managed, we cannot isolate it.

The `Property` is generated automatically, it is not present as usual. That is also the default behavior for `Property`s, and I would not change that. Could you give me a code example what you would expect to work?

**Comment by wolfs** [comment]
@lptr Please take another look!
